src/main/kotlin/com/pubnub/api/retry/RetryableRestCaller.kt

```diff
@@ -25,45 +22,51 @@ internal class RetryableRestCaller<T>(
     internal fun executeRestCallWithRetryPolicy(callToBeExecuted: Call<T>): Response<T> {
         call = callToBeExecuted
         var numberOfAttempts = 0
-        val maxRetryNumber = getRetryCount()
-        val maxAttempts = NUMBER_OF_REGULAR_CALLS + maxRetryNumber
         while (true) {
-            val response = executeRestCall()
-            if (!shouldRetry(response) || numberOfAttempts++ >= maxAttempts) {
-                return response
+            val (response, exception) = executeRestCall()
+            if (!shouldRetry(response) || numberOfAttempts++ >= maxRetryNumberFromConfiguration) {
+                if (response.isSuccessful || isHttpIssueRepresentedByErrorCodeNotByException(exception)) {
```

Comment:
it takes longer to read this and to go to another place to read the implementation, than if you just said `exception == null` here XD

but even more seriously, because of this function call you also have to use `exception!!` below. If you simply inline "exception == null" here, you will get a smart cast below and don't have to use "!!"

strongly suggest removing this function, it makes it more difficult to read what's happening

Reply:
if you want to be explicit when this might happen, I suggest adding a 
```
// comment above the "if" line to explain
if (response...
```

Reply:
I kind of see your point. I also don't like adding "!!" . I see that logic inside the method is simple but my goal was to convey important fact. 
The thing is that it is not that obvious that retry can result from specific exception but also when there is no exception but  response with specific error code. 
I would like to express this.
I most cases(especially in private methods) I prefer to create method name that is descriptive than short method name with comment.

What do you think?